File: Chapter 11 - Mediation and Instrumental Variables/chap11 - analyzing data.py
# -*- coding: utf-8 -*-
"""
Created on Sun May 10 11:44:30 2020

@author: Florent
"""

#################################
##### This script analyzes the data used in chapter 12, 
##### Mediation and Instrumental Variables
#################################

##### Setup #####

#Common libraries
import os
import pandas as pd
from statsmodels.formula.api import ols
import numpy as np


#Loading the data from the chapter on moderation, then setting the right folder
os.chdir('C:\\Users\\Florent\\Dropbox\\Synchronised\\Work_and_projects\\Behavioral data science book\\R scripts\\Part IV Advanced tools\\Chapter 11 - Introduction to moderation')
hist_data_df = pd.read_csv('chap11-historical_data.csv')
os.chdir('C:\\Users\\Florent\\Dropbox\\Synchronised\\Work_and_projects\\Behavioral data science book\\R scripts\\Part IV Advanced tools\\Chapter 12 - Mediation and Instrumental Variables')

##### 1. Mediation #####

#### 1.1. Definition and Measurement ####

ols("duration~play_area", data=hist_data_df).fit().summary()
ols("groceries_purchases~play_area", data=hist_data_df).fit().summary()
ols("groceries_purchases~duration", data=hist_data_df).fit().summary()
ols("groceries_purchases~duration+play_area", data=hist_data_df).fit().summary()

### Bootstrap CI for percentage mediated 

def percentage_mediated_fun(dat_df):
    
    total_effect = ols("groceries_purchases~play_area", data=dat_df).fit(disp=0).params['play_area']
    coeff_med1 = ols("duration~play_area", data=dat_df).fit(disp=0).params['play_area']
    coeff_med2 = ols("groceries_purchases~duration", data=dat_df).fit(disp=0).params['duration']
    mediated_effect = coeff_med1 * coeff_med2
    percentage_mediated = mediated_effect / total_effect
    return percentage_mediated 

percentage_mediated_fun(hist_data_df)

def boot_CI_fun(dat_df, metric_fun, B = 100):
  #Setting sample size
  N = len(dat_df)
  conf_level = 0.9
  coeffs = []
  
  
  for i in range(B):
      sim_data_df = dat_df.sample(n=N, replace = True)
      coeff = metric_fun(sim_data_df)
      coeffs.append(coeff)
  
  coeffs.sort()
  start_idx = round(B * (1 - conf_level) / 2)
  end_idx = - round(B * (1 - conf_level) / 2)
  
  confint = [coeffs[start_idx], coeffs[end_idx]]  
  
  return(confint)
#boot_CI_fun(hist_data_df, percentage_mediated_fun)

#### 1.4. Reducing Uncertainty in Coefficients ####

# NEED MORE WORK HERE

##### 2. Instrumental variables #####

#Common libraries
import os
import pandas as pd
from statsmodels.formula.api import ols
import numpy as np

#Libraries for IV regressions
from linearmodels.iv import IV2SLS

#Loading the data from the chapter 10, then setting the right folder

os.chdir('C:\\Users\\Florent\\Dropbox\\Synchronised\\Work_and_projects\\Behavioral data science book\\R scripts\\Part III Experimental design\\Chapter 10 - cluster randomization and hierarchical modeling')
hist_data_df = pd.read_csv('chap10-historical_data.csv')
exp_data_df = pd.read_csv('chap10-experimental_data.csv')
os.chdir('C:\\Users\\Florent\\Dropbox\\Synchronised\\Work_and_projects\\Behavioral data science book\\R scripts\\Part IV Advanced tools\\Chapter 12 - Mediation and Instrumental Variables')

#Reformat group variable to binary in experimental data
exp_data_df.group = np.where(exp_data_df.group == 'treat', 1, 0)


#Reduced regression, coeff = 1.6
red_mod = ols("M6Spend~group+age+reason", data=exp_data_df).fit(disp=0)
red_mod.summary()

#First stage regression, coeff = 0.5
S1_mod = ols("call_CSAT~group+age+reason", data=exp_data_df).fit(disp=0)
S1_mod.summary()

#Baseline (biased) regression, coeff = 4.00
lm_mod = ols("M6Spend~call_CSAT+age+reason", data=exp_data_df).fit(disp=0)
lm_mod.summary()

#IV regression, coeff = 2.99
iv_mod = IV2SLS.from_formula('M6Spend ~ 1 + age + reason + [call_CSAT ~ group]', 
                             exp_data_df).fit()
iv_mod.params





















hist_data_df['age_quart'] = pd.cut(hist_data_df['age'], 4, 
                                   labels=['q4', 'q3', 'q2', 'q1'],
                                   include_lowest=True)

hist_data_df.groupby('play_area').agg(M = ('duration', lambda x: x.mean()),
    SD = ('duration', lambda x: x.std()))
                                     
#### Multiple moderators ####

# Measure coefficient for moderated moderation
ols("duration~play_area * children * age", data=hist_data_df).fit().summary()

# Determine 90%-CI for moderated moderation coefficient
def metric_fun(dat_df):
    model = ols("duration~play_area * children * age", data=dat_df)
    res = model.fit(disp=0)
    coeff = res.params['play_area:children:age']
    return coeff
#metric_fun(hist_data_df)



#boot_CI_fun(hist_data_df, metric_fun)

#### Validating moderation with Bootstrap ####

#Changing the metric function
def metric_fun2(dat_df):
    model = ols("duration~play_area * children", data=dat_df)
    res = model.fit(disp=0)
    coeff = res.params['play_area:children']
    return coeff
metric_fun(hist_data_df)

# Changing the bootstrap function
def boot_CI_fun2(dat_df, metric_fun, B = 100):
  #Setting the number of bootstrap samples
  
  #Setting sample size
  N = 10000
  conf_level = 0.9
  coeffs = []
  
  
  for i in range(B):
      sim_data_df = dat_df.sample(n=N, replace = True)
      coeff = metric_fun(sim_data_df)
      coeffs.append(coeff)
  
  coeffs.sort()
  start_idx = round(B * (1 - conf_level) / 2)
  end_idx = - round(B * (1 - conf_level) / 2)
  
  confint = [coeffs[start_idx], coeffs[end_idx]]  
  
  return(confint)

boot_CI_fun2(hist_data_df, metric_fun2, B = 1000)

### Interpreting individual coefficients ###

# Centering the age variable in the historical data
centered_data_df = hist_data_df.copy()
centered_data_df['age'] = centered_data_df['age'] \
    .subtract(centered_data_df['age'].mean())

ols("duration~play_area * age", data=hist_data_df).fit().summary()
ols("duration~play_area * age", data=centered_data_df).fit().summary()

# Changing the default for PlayArea to 1
centered_data_df['play_area'] = centered_data_df['play_area']

ols("duration~play_area * age", data=hist_data_df).fit().summary()
ols("duration~play_area * age", data=centered_data_df).fit().summary()


### Calculating effects at the level of business decisions

def business_metric_fun(dat_df):
    
    model =  ols("duration~play_area * (children + age)", data=dat_df)
    res = model.fit(disp=0)
    
    action_dat_df = dat_df[dat_df.play_area == 0].copy()
    action_dat_df['pred_dur0'] = res.predict(action_dat_df) 
    action_dat_df.play_area = 1
    action_dat_df['pred_dur1'] = res.predict(action_dat_df)
    action_dat_df['pred_dur_diff'] = \
        action_dat_df.pred_dur1 - action_dat_df.pred_dur0
    action_res_df = action_dat_df.groupby(['store_id']) \
        .agg(mean_dur_diff=('pred_dur_diff', 'mean'), 
             tot_dur_diff=('pred_dur_diff', 'sum'))
    
    return action_res_df

action_res_df = business_metric_fun(hist_data_df)
action_res_df.describe()

File: Chapter 11 - Mediation and Instrumental Variables/chap11 - analyzing data.R
#################################
##### This script analyzes the data used in chapter 12, 
##### Mediation and Instrumental Variables
#################################

##### Setup #####

#Common libraries
library(tidyverse)
library(ggpubr)
library(boot)
library(rstudioapi)
### Setting the working directory to the parent folder of this script (Rstudio only)
sourceDir <- rstudioapi::getActiveDocumentContext()$path %>% str_extract("^.+/")
setwd(sourceDir)

#Loading the data from the chapter on moderation
hist_data <- read_csv("chap11-historical_data.csv")

#Formatting the data
hist_data <- hist_data %>%
  mutate(store_id = factor(store_id)) %>%
  mutate(day = factor(day)) %>%
  mutate(children = factor(children)) %>%
  mutate(play_area = factor(play_area))

##### 1. Mediation #####

#### 1.1. Definition and Measurement ####

summary(lm(duration~play_area, data=hist_data))
summary(lm(groceries_purchases~play_area, data=hist_data))
summary(lm(groceries_purchases~duration, data=hist_data))
summary(lm(groceries_purchases~duration+play_area, data=hist_data))

### Bootstrap CI for percentage mediated 
percentage_mediated_fun <- function(dat){
  total_effect <- summary(lm(groceries_purchases~play_area, data=dat))$coefficients['play_area1', 'Estimate']
  coeff_med1 <- summary(lm(duration~play_area, data=dat))$coefficients['play_area1', 'Estimate']
  coeff_med2 <- summary(lm(groceries_purchases~duration, data=dat))$coefficients['duration', 'Estimate']
  mediated_effect <- coeff_med1 * coeff_med2
  percentage_mediated <- mediated_effect / total_effect
  return(percentage_mediated)
}
percentage_mediated_fun(hist_data)

boot_CI_fun <- function(dat, metric_fun, B = 100){

  boot_metric_fun <- function(dat, J){
    boot_dat <- dat[J,]
    return(metric_fun(boot_dat))
  }
  boot.out <- boot(data=dat, statistic=boot_metric_fun, R=B)
  confint <- boot.ci(boot.out, conf = 0.90, type = c('perc'))
  CI <- confint$percent[c(4,5)]
  
  return(CI)
}
boot_CI_fun(hist_data, percentage_mediated_fun)

#### 1.4. Reducing Uncertainty in Coefficients ####

#Measuring the overall aggregate effect
agg_effect_fun <- function(dat){
  total_effect <- summary(lm(groceries_purchases~play_area, 
                             data=dat))$coefficients['play_area1', 'Estimate']
  return(total_effect)
}

#Measuring the total effect through its components
composed_effect_fun <- function(dat){
  coeff_med1 <- summary(lm(duration~play_area, 
                           data=dat))$coefficients['play_area1', 'Estimate']
  coeff_med2 <- summary(lm(groceries_purchases~duration, 
                           data=dat))$coefficients['duration', 'Estimate']
  mediated_effect <- coeff_med1 * coeff_med2
  direct_effect <- summary(lm(groceries_purchases~play_area+duration, 
                              data=dat))$coefficients['play_area1', 'Estimate']
  total_effect <- mediated_effect + direct_effect
  return(mediated_effect)
}

#Creating a smaller sample for increased variability
hist_data_sample <- hist_data %>%
  group_by(store_id) %>%
  slice_sample(n=25)

agg_CI <- boot_CI_fun(hist_data_sample, agg_effect_fun)
composed_CI <- boot_CI_fun(hist_data_sample, composed_effect_fun)
agg_CI
composed_CI

##### 2. Instrumental variables #####

#Common libraries
library(tidyverse)
library(ggpubr)
library(boot)

#Libraries for IV regressions
library(ivreg)
library(ivprobit)

#Loading the data from the chapter 10, then setting the right folder
setwd("C:/Users/Florent/Dropbox/Synchronised/Work_and_projects/Behavioral data science book/R scripts/Part III Experimental design/Chapter 10 - cluster randomization and hierarchical modeling")
hist_data <- read_csv("chap10-historical_data.csv")
exp_data <- read_csv("chap10-experimental_data.csv")
setwd("C:/Users/Florent/Dropbox/Synchronised/Work_and_projects/Behavioral data science book/R scripts/Part IV Advanced tools/Chapter 12 - Mediation and Instrumental Variables")

#Reformating the data
hist_data <- hist_data %>%
  mutate(center_ID = factor(center_ID)) %>%
  mutate(rep_ID = factor(rep_ID)) %>%
  mutate(reason = factor(reason))
exp_data <- exp_data %>%
  mutate(center_ID = factor(center_ID, levels = levels(hist_data$center_ID))) %>%
  mutate(rep_ID = factor(rep_ID, levels = levels(hist_data$rep_ID))) %>%
  mutate(reason = factor(reason, levels = levels(hist_data$reason))) %>%
  mutate(group = factor(group)) 

#Reduced regression, coeff = 1.6
red_mod <- lm(M6Spend~group+age, data=exp_data)
summary(red_mod)

#First stage regression, coeff = 0.5
S1_mod <- lm(call_CSAT~group+age, data=exp_data)
summary(S1_mod)

#Baseline (biased) regression
lm_mod <- lm(M6Spend~call_CSAT+age, data=exp_data)
summary(lm_mod)

### IV regression
iv_mod <- ivreg::ivreg(M6Spend~call_CSAT + age | group + age, data=exp_data)
summary(iv_mod)

iv_metric_fun <- function(dat){
  iv_mod <- ivreg::ivreg(M6Spend~call_CSAT + age | group + age, data=dat)
  iv_summ <- summary(iv_mod)
  coeff <- iv_summ$coefficients['call_CSAT','Estimate']
  return(coeff)
}
boot_CI <- boot_CI_fun(exp_data, iv_metric_fun)




# IV regression for binary effect (probit regression)
log_data <- exp_data %>%
  mutate(spend_bin = ifelse(M6Spend >= 50, 1, 0)) %>%
  #mutate(spend_bin = factor(spend_bin, levels=c(0,1))) %>%
  select(group, spend_bin, call_CSAT, age)

#summary(glm(spend_bin~call_CSAT+age, data=log_data, family = binomial(link = "logit")))
summary(glm(spend_bin~call_CSAT+age, data=log_data, family = binomial(link = "probit")))
summary(glm(spend_bin~group, data=log_data, family = binomial(link = "probit")))
summary(glm(spend_bin~call_CSAT, data=log_data, family = binomial(link = "probit")))

summary(ivprobit(spend_bin~age|call_CSAT|group, data=log_data))
